worker/provisioner/provisioner_task.go

```diff
@@ -689,106 +714,357 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
 	return nil
 }
 
-func (task *provisionerTask) startMachines(machines []*apiprovisioner.Machine) error {
-	for _, m := range machines {
-		// Make sure we shouldn't be stopping before we start the next machine
-		select {
-		case <-task.catacomb.Dying():
-			return task.catacomb.ErrDying()
-		default:
-		}
+type availabilityZoneMachine struct {
+	zoneName         string
+	machineIds       set.Strings
+	failedMachineIds set.Strings
+}
+
+// populateAvailabilityZoneMachines fills in the map, availabilityZoneMachines,
+// if empty, with a current mapping of availability zone to IDs of machines
+// running in that zone.  If it's not empty, check to see if we should mark
```

Comment:
The "If it's not empty..." bit doesn't apply anymore?

Reply:
updated